stablehlo/reference/Ops.cpp

```diff
@@ -869,50 +893,39 @@ Tensor evalGatherOp(const Tensor &operand, const Tensor &startIndices,
                     ShapedType resultType) {
   Tensor result(resultType);
   Axes batchDims;
-  for (auto i = 0; i < result.getRank(); ++i)
-    if (!llvm::is_contained(offsetDims, i)) batchDims.push_back(i);
+  for (auto d : result.getAxes())
+    if (!llvm::is_contained(offsetDims, d)) batchDims.push_back(d);
 
   for (auto resultIt = result.index_begin(); resultIt != result.index_end();
        ++resultIt) {
     Index batchIndex;
-    for (auto batchDim : batchDims) batchIndex.push_back((*resultIt)[batchDim]);
-
-    Index sliceStartIndices(batchIndex);
-    Index sliceLimitIndices(Sizes(sliceStartIndices) + 1);
-    if (indexVectorDim < startIndices.getRank()) {
-      sliceStartIndices.insert(sliceStartIndices.begin() + indexVectorDim, 0);
-      sliceLimitIndices.insert(sliceLimitIndices.begin() + indexVectorDim,
-                               startIndices.getShape()[indexVectorDim]);
-    }
-    auto startIndexSlice =
-        evalSliceOp(startIndices, sliceStartIndices, sliceLimitIndices,
-                    Sizes(startIndices.getRank(), 1));
-    Index startIndex;
-    for (auto it = startIndexSlice.index_begin();
-         it != startIndexSlice.index_end(); ++it)
-      startIndex.push_back(
-          startIndexSlice.get(*it).getIntegerValue().getSExtValue());
+    for (auto d : batchDims) batchIndex.push_back((*resultIt)[d]);
+
+    Index startIndicesIndex = batchIndex;
+    if (indexVectorDim < startIndices.getRank())
+      startIndicesIndex.insert(startIndicesIndex.begin() + indexVectorDim,
+                               kColon);
+    Index startIndex = evalIndex(evalSliceOp(startIndices, startIndicesIndex));
```

Comment:
Having to use the unmodified `evalSliceOp` was seriously confusing to me (and to the previous reviewer as well), so I took a stab at introducing a notation that closely follows the spec.